### ISSUE TITLE
libxsmm_config.h not found when building plaidml-v1 from source

### ISSUE THREAD

**Issue by powderluv** [plaidml/plaidml#973]
INFO: From Compiling external/llvm-project/mlir/lib/Dialect/SPIRV/SPIRVCanonicalization.cpp:
external/llvm-project/mlir/lib/Dialect/SPIRV/SPIRVCanonicalization.cpp:272:1: warning: multi-line comment [-Wcomment]
 //                            /   \
 ^
ERROR: /home/foo/github/plaidml/pmlc/rt/BUILD:5:1: C++ compilation of rule '//pmlc/rt:rt' failed (Exit 1)
In file included from external/xsmm/include/libxsmm_typedefs.h:34,
                 from external/xsmm/include/libxsmm_generator.h:34,
                 from external/xsmm/include/../src/generator_common.h:34,
                 from external/xsmm/include/../src/generator_common.c:31,
                 from external/xsmm/include/libxsmm_source.h:55,
                 from pmlc/rt/xsmm.cc:3:
external/xsmm/include/libxsmm_macros.h:34:10: fatal error: libxsmm_config.h: No such file or directory
 #include "libxsmm_config.h"
          ^~~~~~~~~~~~~~~~~~
compilation terminated.
Target //plaidml:wheel failed to build
Use --verbose_failures to see the command lines of failed build steps.

**Comment by flaub** [comment]
We've not seen any of these issues, I'm wondering if you're following these instructions:
https://plaidml.readthedocs.io/en/plaidml-v1/usage/building.html

From what I can gather, the `xsmm_repo` custom repository rule didn't run.

BTW, we have CI for all major platforms and we don't see this issue: 
https://buildkite.com/bazel/plaidml/builds?branch=plaidml-v1

**Comment by powderluv** [comment]
I am following those instructions. I just tested on a 18.04 system and it passed. Earlier I was trying on the  20.04 LTS system (python doesn't exist there and you will have to refer to python2 or python3)

I'll try to debug further and report with specific issues.

**Comment by hfp** [comment]
Indeed, LIBXSMM's [configure script](https://github.com/hfp/libxsmm/blob/master/scripts/libxsmm_config.py#L1) refers to "python" (`#!/usr/bin/env python`). If the OS distribution does not provide "python" (as opposed to python2 or python3), the script may fail.

( LIBXSMM's own build system (Make) takes the above case into account by creating a symlink in the current working directory (during Make) pointing for instance to "python3". )

PlaidML uses Bazel and does not rely on LIBXSMM's Makefile, hence the problem may be solved by PlaidML (this assumes that `libxsmm_config.py` is still leveraged). Let me know if a change in LIBXSMM can help.

**Comment by powderluv** [comment]
yes that will be very useful since "python" will disappear in Ubuntu 20.04 LTS next month when it is released. Thx

**Comment by hfp** [comment]
I wonder if your Ubuntu 20 LTS misses some configuration. Typically, most distributions allow to call whatever Python versions is preferred to be called with "python" (still holds true if Python v2 disappears). Perhaps it's something like `update-alternatives --config python`, but I am not sure. It could be also an artifact of an (incomplete) upgrade.

**Comment by powderluv** [comment]
"The latest word from Matthias Klose, the Debian developer who has been coordinating much of the Python 2 removal effort, is that libpython-dev, python, python-minimal, and other packages no longer exist as part of the python-defaults package in the Ubuntu 20.04 archive."

https://www.fosslinux.com/26413/ubuntu-20-04-lts-focal-fossa-set-to-remove-python-2.htm

**Comment by hfp** [comment]
@powderluv I think this is misinterpreted, there is no point that an executable called `python` must be Python2. Removing a package called "python" (referring to Python2) is only losely related to an executable called `python`. Though, you also mentioned the package "python-minimal" but you may not expect an executable called `python-minimal`. If your interpretation would be right, we must use `python4` in some (far) future? I still think this can be resolved with `update-alternatives`.

**Comment by flaub** [comment]
The intent of bazel is to make all build actions hermetic. Therefore, in principle, it shouldn't matter what version of python is installed on the system, or even if python is installed at all. Clearly in this case the build action is 'leaking' and somehow isn't working the same way as our CI build environment. I consider this to be a bug in our build rules rather than having others do something to change their build environment.

This is an attempt to make things more hermetic: https://github.com/plaidml/plaidml/pull/988
I'm not sure that the above PR will address this particular issue, but it'd be interesting to know if it does.

@hfp Actually what would be useful for us is if there was a single python script that I can call that will setup the entire project (generating header files, etc). Right now we're calling the `make header-only`, but this isn't cross-platform safe (what to do on Windows?) and also breaks hermicity for us.

**Comment by hfp** [comment]
@flaub Thanks for the feedback!  I took steps to get almost rid of configuration especially wrt Windows. I have to document this including the persective of 3rd party build systems that want minimal effort to get things completed (without hard-coding insight into our library).

**Comment by powderluv** [comment]
I tried PR #988 but that doesn't help with libxsmm.h

ERROR: /home/foo/github/plaidml/pmlc/rt/BUILD:6:1: C++ compilation of rule '//pmlc/rt:rt' failed (Exit 1)
In file included from external/xsmm/include/../src/generator_common.c:32, 
                 from external/xsmm/include/libxsmm_source.h:55,
                 from pmlc/rt/xsmm.cc:4:                             
external/xsmm/include/../src/libxsmm_main.h:34:10: fatal error: libxsmm.h: No such file or directory
 #include <libxsmm.h>                  
          ^~~~~~~~~~~
compilation terminated.                                                                                                                           
Target //plaidml:wheel failed to build
INFO: Elapsed time: 633.921s, Critical Path: 123.14s           
INFO: 1544 processes: 1544 local.
FAILED: Build did NOT complete successfully    
     
1 foo@foo:~/github/plaidml$ git log
commit bb957bfedee2cb6121bf5565b641ac1735e3c4c6 (HEAD -> flaub-hermit, origin/flaub-hermit)
Author: Frank Laub <frank.laub@intel.com>      
Date:   Sun Mar 29 23:20:59 2020 +0000

    Merge unix + windows conda repos

**Comment by hfp** [comment]
LIBXSMM zero-config:
* [https://libxsmm.readthedocs.io/#zero-config-abi](https://libxsmm.readthedocs.io/#zero-config-abi)
* [https://libxsmm.readthedocs.io/#zero-config](https://libxsmm.readthedocs.io/#zero-config)

**Comment by powderluv** [comment]
sorry what should I do with zero-config ? 

bazelisk build --verbose_failures //plaidml:wheel

Top of master this morning fails with :

external/llvm-project/llvm/lib/Support/APFloat.cpp:2697:53: warning: 'D.llvm::decimalInfo::firstSigDigit' may be used uninitialized in this function [-Wmaybe-uninitialized]
   if (D.firstSigDigit == str.end() || decDigitValue(*D.firstSigDigit) >= 10U) {
                                                     ^~~~~~~~~~~~~~~~
ERROR: /home/foo/github/plaidml/pmlc/rt/BUILD:6:1: C++ compilation of rule '//pmlc/rt:rt' failed (Exit 1)
In file included from external/xsmm/include/../src/generator_common.c:32,
                 from external/xsmm/include/libxsmm_source.h:55,
                 from pmlc/rt/xsmm.cc:4:
external/xsmm/include/../src/libxsmm_main.h:34:10: fatal error: libxsmm.h: No such file or directory
 #include <libxsmm.h>
          ^~~~~~~~~~~
compilation terminated.
Target //plaidml:wheel failed to build
INFO: Elapsed time: 65.283s, Critical Path: 9.37s
INFO: 202 processes: 202 local.
FAILED: Build did NOT complete successfully

**Comment by hfp** [comment]
@powderluv My comment was a note for @flaub regarding the latter paragraph of his [comment](https://github.com/plaidml/plaidml/issues/973#issuecomment-605713282). Sorry, I was not clear that this is nothing actionable for you at this point. Frank may pickup this work and perhaps can help resolving the original issue you reported. Sorry again for hooking up your issue report.

**Comment by flaub** [comment]
@hfp Thanks we'll take a look at zero-config and try to use it in a future update :)

**Comment by hfp** [comment]
@powderluv I have updated LIBXSMM to prefer Python3 interpreter (if available). Further, I have updated all Python scripts to carry `#!/usr/bin/env python3` (during LIBXSMM's `make` the Shebang does not matter since the correct Python interpreter is prepended). With this, even PlaidML's current integration should work "as is" (no need to incorporate the new new zero-config feature). For the time being, one needs to rely on the unreleased/master version of LIBXSMM.

Here is what I did (edit `bzl/workspace.bzl`):

```text
    xsmm_repo(
        name = "xsmm",
        url = "https://github.com/hfp/libxsmm/archive/1e02a56102d215f00dd5cf810ad28d6d5dcc22ad.zip",
        sha256 = "c6abfb0842438cbafba7a94374e0f32e6131d18eafe3445ff469890ca8a16aa7",
        strip_prefix = "libxsmm-1e02a56102d215f00dd5cf810ad28d6d5dcc22ad",
        build_file = clean_dep("//bzl:xsmm.BUILD"),
    )
```

I have to admit, I tested on Debian and not bleeding edge Ubuntu. However, everything should work there as well. I am confident this does not wastes your time.

**Comment by hfp** [comment]
For plaidml-v1, above `build_file` must be `build_file = clean_dep("//vendor/xsmm:xsmm.BUILD"),`.

**Comment by powderluv** [comment]
I can confirm I can build with your change. Thanks

**Comment by alheinecke** [comment]
I have submitted a PR for this #1023 ; this bumps LIBXSMM to the latest master version which also has much better small GEMM performance and support for batch reduce GEMM which we are planning to integrate into PlaidML: https://arxiv.org/abs/1906.06440

**Comment by flaub** [comment]
@hfp I'm trying to use the latest zero-config approach, and so far it works great on all platforms, except for Windows. I see this whenever I try to run a test that uses xsmm:

```
LIBXSMM ERROR: application must be linked against LAPACK/BLAS sgemm!
```

Everything builds fine, this only happens at runtime. I've attempted to use the following #define values (and every combination) to no avail:

```
BLAS=0
__BLAS=0
LIBXSMM_NO_BLAS=1
LIBXSMM_BLAS=0
LIBXSMM_DEFAULT_CONFIG
```

Any ideas?

**Comment by hfp** [comment]
@flaub Let me fix this, thanks for reporting!

Regarding Windows calling convention, @alheinecke may have an update.

**Comment by hfp** [comment]
@flaub Can you share your changes to use zero-config? Perhaps a branch on your fork of PlaidML...

**Comment by flaub** [comment]
@hfp It's part of this PR: https://github.com/plaidml/plaidml/pull/1025

**Comment by hfp** [comment]
@flaub  Thanks!  I will get this reproduced then under Windows. I will let you know about the fix.
